### PR TITLE
Add HTTP multiplexing

### PR DESCRIPTION
**Pull request by delner**
As a PR that replaces #882 (incorporating @stormsilver's original contributions; thank you!), this pull request aims to add "multiplexing" to the other HTTP client libraries instrumented.

In practice, "multiplexing" allows users to configure different tracer settings for different URLs. The most common reason why you might prefer to do this is to mark some outgoing HTTP calls as belonging to one service, while marking others as belonging to another.

e.g.

```ruby
Datadog.configure do |c|
 # Assign all outgoing Net::HTTP requests to amazonaws.com to the 'aws' service
  c.use :http, describe: /amazonaws\.com/ do |http|
    http.service_name = 'aws'
  end
end
```

It also adds `split_by_domain` options which assigns the FQDN of the request URL as the service name.

These changes affect the following libraries:
 - Ethon
 - Excon
 - Faraday
 - HTTP

### CODE REVIEW

**Comment by delner** [comment]
Changes are working, but might want to do a little refactoring/testing before merging. Could use a review from you as well @marcotc.

File: lib/ddtrace/contrib/configuration/resolvers/regexp_resolver.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      module Resolvers
+        # Matches strings against Regexps.
+        class RegexpResolver < Datadog::Contrib::Configuration::Resolver
+          def resolve(name)
+            # Co-erce to string
+            name = name.to_s
+
+            # Try to find a matching pattern
+            matching_pattern = patterns.find do |pattern|
+              pattern.is_a?(Regexp) ? pattern =~ name : pattern == name
```
Comment (marcotc) [inline lib/ddtrace/contrib/configuration/resolvers/regexp_resolver.rb]: I have a suggestion we can discuss for this matching: how about using case equality `#===` here?
It supports regular expressions, simple string matches, and procs (`pattern === name`). The logic would look simpler too.

I think procs might be something users might be interested in using for matching patterns here.

Comment (delner) [reply lib/ddtrace/contrib/configuration/resolvers/regexp_resolver.rb]: Okay, I'll give case equality a shot.

Using procs could be good, too, although there's a small risk that in allowing that could end up with users introducing checks that have a high performance cost (as this is run per span, effectively.) Benefits probably outweigh this risk though, so I think its a good suggestion.

Comment (delner) [reply lib/ddtrace/contrib/configuration/resolvers/regexp_resolver.rb]: Case equality seems to work nicely; also implemented Proc support. Thanks for the suggestion!

**Pull request merged**